File: backend/src/main/java/com/github/aadvorak/artilleryonline/battle/calculator/DroneTargetCalculator.java
package com.github.aadvorak.artilleryonline.battle.calculator;

import com.github.aadvorak.artilleryonline.battle.calculations.BattleCalculations;
import com.github.aadvorak.artilleryonline.battle.calculations.DroneCalculations;
import com.github.aadvorak.artilleryonline.battle.common.Position;
import com.github.aadvorak.artilleryonline.battle.utils.GeometryUtils;

import java.util.Set;
import java.util.stream.Collectors;

public class DroneTargetCalculator {

    public static void calculate(DroneCalculations drone, BattleCalculations battle) {
        if (drone.getModel().getState().isDestroyed()) {
            drone.getModel().getState().getGunState().setTriggerPushed(false);
            return;
        }
        var ammo = drone.getModel().getState().getAmmo().values().iterator().next();
        var targets = getTargets(drone, battle, ammo);
        if (targets.isEmpty()) {
            return;
        }
        var dronePosition = drone.getModel().getState().getPosition().getCenter();
        var xDiffMap = targets.stream()
                .collect(Collectors.toMap(
                        position -> position.getX() - dronePosition.getX(),
                        position -> position));
        var iterator = xDiffMap.keySet().iterator();
        var minXDiff = iterator.next();
        while (iterator.hasNext()) {
            var xDiff = iterator.next();
            if (Math.abs(xDiff) < Math.abs(minXDiff)) {
                minXDiff = xDiff;
            }
        }
        var targetPosition = xDiffMap.get(minXDiff);
        var gunAngle = drone.getModel().getState().getGunAngle() + drone.getModel().getState().getPosition().getAngle();
        var angleDiff = GeometryUtils.calculateAngleDiff(gunAngle, dronePosition.angleTo(targetPosition));
        drone.getModel().getState().getGunState().setTriggerPushed(Math.abs(angleDiff) < Math.PI / 32
                && dronePosition.distanceTo(targetPosition) < drone.getModel().getSpecs().getFlyHeight());
        drone.setTarget(new DroneCalculations.Target()
                .setXDiff(minXDiff)
                .setAngleDiff(angleDiff)
        );
    }

    private static Set<Position> getTargets(DroneCalculations drone, BattleCalculations battle, Integer ammo) {
        if (ammo > 0) {
            return TargetCalculator.calculatePositions(drone.getModel().getVehicleId(), battle);
        } else {
            return battle.getModel().getVehicles().values().stream()
                    .filter(vehicleModel -> drone.getModel().getVehicleId() != null
                            && vehicleModel.getId() == drone.getModel().getVehicleId())
                    .map(vehicleModel -> vehicleModel.getState().getPosition().getCenter())
                    .collect(Collectors.toSet());
        }
    }
}

File: frontend/playground/battle/calculator/drone-target-calculator.ts
import type {DroneCalculations} from "~/playground/data/calculations";
import type {BattleModel} from "~/playground/data/model";
import type {Position} from "~/playground/data/common";
import {VectorUtils} from "~/playground/utils/vector-utils";
import {BattleUtils} from "~/playground/utils/battle-utils";
import {TargetCalculator} from "~/playground/battle/calculator/target-calculator";

export const DroneTargetCalculator = {
  calculate(drone: DroneCalculations, battleModel: BattleModel) {
    if (drone.model.state.destroyed) return

    const ammo = Object.values(drone.model.state.ammo)[0]
    let targets: Position[] = []

    if (ammo && ammo > 0) {
      targets = TargetCalculator.calculatePositions(drone.model.vehicleId, battleModel)
    } else {
      targets = Object.values(battleModel.vehicles)
          .filter(vehicle => vehicle.id === drone.model.vehicleId)
          .map(vehicle => ({
            x: vehicle.state.position.x,
            y: vehicle.state.position.y
          }))
    }

    if (targets.length === 0) return

    const dronePosition = drone.model.state.position
    const xDiffMap: { [key: number]: Position } = {}

    targets.forEach(position => {
      const xDiff = position.x - dronePosition.x
      xDiffMap[xDiff] = position
    })

    let minXDiff = Math.min(...Object.keys(xDiffMap).map(Number))

    const targetPosition = xDiffMap[minXDiff]!
    const gunAngle = drone.model.state.gunAngle + drone.model.state.position.angle
    const angleDiff = BattleUtils.calculateAngleDiff(gunAngle, VectorUtils.angleFromTo(dronePosition, targetPosition))

    drone.target = {
      xDiff: minXDiff,
      angleDiff: angleDiff
    }
  }
}
